---
title: "Untitled"
author: "avanvalken"
date: "2024-12-04"
output: html_document
---

```{r setup, include=FALSE}
library(tidyverse)
library(scRepertoire)
library(immunarch)
knitr::opts_chunk$set(echo = TRUE)
```



# Load Data
```{r}
#setwd("..")
# path to contigs
path <- file.path(getwd(), "data/trust4_reports")
contig.output <- list.files(path, pattern = "airr.tsv", recursive = T, full.names = T)

# read in contigs
contig.list <- lapply(contig.output, read_tsv)
names(contig.list) <- list.dirs(path, full.names = F, recursive = F)

contig.list <- loadContigs(contig.list, format="AIRR")
```

# scRepertoire

Citation is: Borcherding, N, Yang, Q, & Safina, K. (2024). scRepertoire v2: A toolkit for single-cell immune receptor profiling. DOI: 10.18129/B9.bioc.scRepertoire



# Immunarch

## load data
```{r}
# path to trust4 reports
path <- file.path("../data/trust4_reports")
immdata <- repLoad(path)

```

## add metadata
```{r}
immdata$meta <- immdata$meta %>% mutate(Status=ifelse(grepl("CC",Sample), "CC", "ZA"))
```


## Explore Repertoire
```{r}
exp_vol <- repExplore(immdata$data, .method = "volume")
exp_clone <- repExplore(immdata$data, .method = "clones")
exp_len <- repExplore(immdata$data, .method = "len", .col="aa")
exp_count <- repExplore(immdata$data, .method = "count")

p1 <- vis(exp_vol, .by = c("Status"), .meta = immdata$meta)
p2 <- vis(exp_clone, .by = c("Status"), .meta = immdata$meta)
p3 <- vis(exp_len, .by = c("Status"), .meta = immdata$meta)
p4 <- vis(exp_count, .by = c("Status"), .meta = immdata$meta)
p1 | p2 
p3 | p4


```

### Save
```{r}
outs <- file.path("../outs/immunarch", "repexplore")
#dir.create(outs)
ggsave(file.path(outs, "unique_clonotypes_by_status_volume.png"), plot=p1, width=5, height=5)
ggsave(file.path(outs, "unique_clonotypes_by_status_clones.png"), plot=p2, width=5, height=5)
ggsave(file.path(outs, "unique_clonotypes_by_status_length.png"), plot=p3, width=5, height=5)
ggsave(file.path(outs, "unique_clonotypes_by_status_count.png"), plot=p4, width=5, height=5)



```



## Clonality
```{r}
outs <- file.path("../outs/immunarch", "clonality")
#dir.create(outs)
```

### proportion of clones
```{r}
# proportion of clones
imm_pr <- repClonality(immdata$data, .method = "clonal.prop")
imm_pr
```
#### Visualize and save
```{r}
vis(imm_pr) + vis(imm_pr, .by = "Status", .meta = immdata$meta)
ggsave(file.path(outs, "proportion.png"), width=10, height = 7)
```

### Top clonaltypes
```{r}
imm_top <- repClonality(immdata$data, .method = "top", .head = c(1, 2, 5, 10, 50, 60))
imm_top
```
#### Visualize and save
```{r}
vis(imm_top) + vis(imm_top, .by = "Status", .meta = immdata$meta)
ggsave(file.path(outs, "top_proportion.png"), width=10, height = 7)
```

### Lowest clonaltypes
```{r}
imm_rare <- repClonality(immdata$data, .method = "rare")
imm_rare
```
#### Visualize and save
```{r}
vis(imm_rare) + vis(imm_rare, .by = "Status", .meta = immdata$meta)
ggsave(file.path(outs, "rare_proportion.png"), width=10, height = 7)
```

###  Homeostasis
Finally, the homeo method assesses the clonal space homeostasis, i.e., the proportion of the repertoire occupied by the clones of a given size:

```{r}
imm_hom <- repClonality(immdata$data,
  .method = "homeo",
  .clone.types = c(Small = .0001, Medium = .001, Large = .01, Hyperexpanded = 1)
)
imm_hom
```

#### Visualize and save
```{r}
vis(imm_hom) + vis(imm_hom, .by = "Status", .meta = immdata$meta)
ggsave(file.path(outs, "homeostasis.png"), width=10, height = 7)
```

## Repertoire overlap
Repertoire overlap is the most common approach to measure repertoire similarity. It is achieved by computation of specific statistics on clonotypes shared between given repertoires, also called “public” clonotypes. immunarch provides several indices: - number of public clonotypes (.method = "public") - a classic measure of overlap similarity.

overlap coefficient (.method = "overlap") - a normalised measure of overlap similarity. It is defined as the size of the intersection divided by the smaller of the size of the two sets.

Jaccard index (.method = "jaccard") - measures the similarity between finite sample sets, and is defined as the size of the intersection divided by the size of the union of the sample sets.

Tversky index (.method = "tversky") - an asymmetric similarity measure on sets that compares a variant to a prototype. If using default arguments, it’s similar to Dice’s coefficient.

cosine similarity (.method = "cosine") - a measure of similarity between two non-zero vectors

Morisita’s overlap index (.method = "morisita") - a statistical measure of dispersion of individuals in a population. It is used to compare overlap among samples.

incremental overlap - overlaps of the N most abundant clonotypes with incrementally growing N (.method = "inc+METHOD", e.g., "inc+public" or "inc+morisita").

The function that includes described methods is repOverlap. Again, the output is easily visualised when passed to vis() function that does all the work:
```{r}
outs <- file.path("../outs/immunarch", "overlap")
dir.create(outs)
```


### Overlap, public and morisita
overlap of public clonotypes; it is defined as the size of the intersection divided by the smaller of the size of the two sets
```{r}
imm_ov1 <- repOverlap(immdata$data, .method = "public", .verbose = F)
imm_ov2 <- repOverlap(immdata$data, .method = "morisita", .verbose = F)

p1 <- vis(imm_ov1)
p2 <- vis(imm_ov2, .text.size = 2)

p1 + p2
```
### Save
```{r}
ggsave(file.path(outs, "public_overlap.png"),plot=p1, width=5, height=5)
ggsave(file.path(outs, "morisita_overlap.png"),plot=p2, width=5, height=5)

```


```{r}
p3 <- vis(imm_ov1, "heatmap2")
```
```{r}
ggsave(file.path(outs, "public_overlap_heatmap.png"), plot=p3, width=5, height=5)
```


```{r}
# Apply different analysis algorithms to the matrix of public clonotypes:
# "mds" - Multi-dimensional Scaling
#repOverlapAnalysis(imm_ov1, "mds") Error in MASS::isoMDS(.data, k = 2, trace = FALSE) : 
  #zero or negative distance between objects 1 and 5
p4 <- repOverlapAnalysis(imm_ov1, "tsne") %>% vis()
p4
```
### Save
```{r}
ggsave(file.path(outs, "public_overlap_tsn3.png"), plot=p4, width=5, height=5)
```

```{r}
#repOverlapAnalysis(imm_ov1, "mds+kmeans") %>% vis()
#Error in MASS::isoMDS(.data, k = 2, trace = FALSE) : 
#  zero or negative distance between objects 1 and 5
```

### Public repertoire
```{r}
# Pass "nt" as the second parameter to build the public repertoire table using CDR3 nucleotide sequences
pr.nt <- pubRep(immdata$data, "nt", .verbose = F)
pr.nt
```

```{r}
# Pass "aa+v" as the second parameter to build the public repertoire table using CDR3 aminoacid sequences and V alleles
# In order to use only CDR3 aminoacid sequences, just pass "aa"
pr.aav <- pubRep(immdata$data, "aa+v", .verbose = F)
pr.aav
```



```{r}
# Create a public repertoire with coding-only sequences using both CDR3 amino acid sequences and V genes
pr <- pubRep(immdata$data, "aa+v", .coding = T, .verbose = F)

# Apply the filter subroutine to leave clonotypes presented only in healthy individuals
pr1 <- pubRepFilter(pr, immdata$meta, c(Status = "ZA"))

# Apply the filter subroutine to leave clonotypes presented only in diseased individuals
pr2 <- pubRepFilter(pr, immdata$meta, c(Status = "CC"))

# Divide one by another
pr3 <- pubRepApply(pr1, pr2)

# Plot it
p <- ggplot() +
  geom_jitter(aes(x = "Treatment", y = Result), data = pr3)
p
```

## Gene usage computation
immunarch comes with a gene segments data table containing known gene segments for several species following the IMGT nomenclature. In order to get the current statistics of genes, call the gene_stats() function:

```{r}
outs <- file.path("../outs/immunarch/gene_usage")
dir.create(outs)
```


### TCR


```{r}
imm_gu <- geneUsage(immdata$data, "hs.trbv", .norm = T)

vis(imm_gu, .by = "Status", .meta = immdata$meta)
```
#### Save
```{r}
ggsave(file.path(outs, "V_gene_all_barplots.png"), width=10, height=7)
```

### Boxplot of V gene use
```{r}
vis(imm_gu, .by = "Status", .meta = immdata$meta, .plot = "box")
```
#### Save
```{r}
ggsave(file.path(outs, "V_gene_all_boxplots.png"), width=10, height=7)
```



## Gene Usage Analysis
```{r}
imm_gu <- geneUsage(immdata$data, "hs.trbv", .norm = T)

imm_gu_js <- geneUsageAnalysis(imm_gu, .method = "js", .verbose = F)
imm_gu_cor <- geneUsageAnalysis(imm_gu, .method = "cor", .verbose = F)

p1 <- vis(imm_gu_js, .title = "Gene usage JS-divergence", .leg.title = "JS", .text.size = 1.5)
p2 <- vis(imm_gu_cor, .title = "Gene usage correlation", .leg.title = "Cor", .text.size = 1.5)

p1 + p2
```
#### Save
```{r}
ggsave(file.path(outs, "JS_divergence_Correlation_plots.png"), width=10, height=7)
```

```{r}
imm_gu_js[is.na(imm_gu_js)] <- 0

vis(geneUsageAnalysis(imm_gu, "cosine+hclust", .verbose = F))
```
#### Save
```{r}
ggsave(file.path(outs, "JS_divergence_cosine_hclust_plots.png"), width=10, height=7)
```

```{r}
imm_cl_pca <- geneUsageAnalysis(imm_gu, "js+pca+kmeans", .verbose = F)
imm_cl_mds <- geneUsageAnalysis(imm_gu, "js+mds+kmeans", .verbose = F)
imm_cl_tsne <- geneUsageAnalysis(imm_gu, "js+tsne+kmeans", .perp = .01, .verbose = F)

p1 <- vis(imm_cl_pca, .plot = "clust")
p2 <- vis(imm_cl_mds, .plot = "clust")
p3 <- vis(imm_cl_tsne, .plot = "clust")
p1 + p2 + p3
```

### Save
```{r}
ggsave(file.path(outs, "JS_PCA_Kmeans.png"), plot=p1,width=5, height=5)
```

```{r}
p1 <- vis(spectratype(immdata$data[[1]], .quant = "id", .col = "nt"))
p2 <- vis(spectratype(immdata$data[[1]], .quant = "count", .col = "aa+v"))

p1 + p2
```
### Save
```{r}
ggsave(file.path(outs, "CDR3nt_AA_V_length.png"), plot=p1,width=10, height=7)
```

## Diversity estimation
```{r}
outs <- file.path("../outs/immunarch/diversity_estimation")
#dir.create(outs)
```

There are several approaches to the estimation of repertoire diversity implemented in the repDiversity function. The .method parameter similarly to above mentioned functions sets the means for diversity estimation. You can choose one of the following methods:

chao1 - is a nonparameteric asymptotic estimator of species richness (number of species in a population).

hill - Hill numbers are a mathematically unified family of diversity indices (differing only by an exponent q).

div - true diversity, or the effective number of types, refers to the number of equally abundant types needed for the average proportional abundance of the types to equal that observed in the dataset of interest where all types may not be equally abundant.

gini.simp - The Gini-Simpson index is the probability of interspecific encounter, i.e., probability that two entities represent different types.

inv.simp - Inverse Simpson index is the effective number of types that is obtained when the weighted arithmetic mean is used to quantify average proportional abundance of types in the dataset of interest.

gini - The Gini coefficient measures the inequality among values of a frequency distribution (for example levels of income). A Gini coefficient of zero expresses perfect equality, where all values are the same (for example, where everyone has the same income). A Gini coefficient of one (or 100 percents ) expresses maximal inequality among values (for example where only one person has all the income).

raref - Rarefaction is a technique to assess species richness from the results of sampling through extrapolation.

The .col parameter regulates what sequences and gene segments to choose. For example, if you want to estimate diversity on the nucleotide level, you need to supply .col = "nt", in case you want to estimate the diversity on the amino acid level - .col = "aa". If you want to estimate diversity of the amino acid CDR3 sequences coupled with V gene segments, you need to provide .col = "aa+v". By default .col = "aa".

### diversity measures
```{r}
# Compute statistics and visualise them
# Chao1 diversity measure
div_chao <- repDiversity(immdata$data, "chao1")

# Hill numbers
div_hill <- repDiversity(immdata$data, "hill")

# D50
div_d50 <- repDiversity(immdata$data, "d50")

# Ecological diversity measure
div_div <- repDiversity(immdata$data, "div")

# Gini-Simpson
div_ginisimp <- repDiversity(immdata$data, "gini.simp")

# Inverse Simpson inv.simp
div_invsimp <- repDiversity(immdata$data, "inv.simp")

# Gini Coefficient
div_gini <- repDiversity(immdata$data, "gini")

p1 <- vis(div_chao)
p2 <- vis(div_chao, .by = c("Status"), .meta = immdata$meta)
p3 <- vis(div_hill, .by = c("Status"), .meta = immdata$meta)

p4 <- vis(div_d50)
p5 <- vis(div_d50, .by = "Status", .meta = immdata$meta)
p6 <- vis(div_div)

p7 <- vis(div_ginisimp, .by = c("Status"), .meta = immdata$meta)
p8 <- vis(div_invsimp, .by = c("Status"), .meta = immdata$meta)
#p9 <- vis(div_gini, .by = c("Status"), .meta = immdata$meta)

p1 + p2
```
### Save
```{r}
ggsave(file.path(outs, "chao1_diversity.png"), plot=p1,width=10, height=7)
```

### Hill and ecological
```{r}
p3 + p6
```
```{r}
ggsave(file.path(outs, "hill_true_diversity.png"), plot=p1,width=10, height=7)
```


### Gini-Simpson and Inverse-Simpson index
```{r}
p7 + p8
```
```{r}
ggsave(file.path(outs, "gini-simpson_inv-simpson_diversity.png"), plot=p1,width=10, height=7)
```


```{r}
p4 + p5
```
```{r}
ggsave(file.path(outs, "div_top50.png"), plot=p1,width=10, height=7)

```


```{r}
imm_raref <- repDiversity(immdata$data, "raref", .verbose = F)

p1 <- vis(imm_raref)
p2 <- vis(imm_raref, .by = "Status", .meta = immdata$meta)

p1 + p2 # need to remove outlier
```


```{r}
repDiversity(immdata$data, "raref", .verbose = F) %>% vis(.log = TRUE,.by = "Status", .meta = immdata$meta)

```

```{r}
ggsave(file.path(outs, "rarefication_by_status_logscale.png"), plot=p1,width=7, height=7)

```



